File: ios/Runner/BlurThresholdPolicy.swift
import Foundation

/// Three-band classification of a sharpness score (variance of Laplacian from
/// `BlurMetric`) into an actionable band — a Swift port of the Android
/// `BlurThresholdPolicy`, so the wire form and band boundaries match exactly.
enum BlurBand: String {
  /// Too blurry — don't use the frame.
  case reject
  /// Borderline — usable but flagged; the capture flow decides.
  case warn
  /// Sharp enough — use the frame.
  case accept

  /// Lowercase wire form for the channel payload (`reject`/`warn`/`accept`).
  var wire: String { rawValue }
}

/// Holds validated thresholds and classifies scores against them (value type —
/// replaced wholesale on a threshold update, so reads are consistent).
struct BlurThresholdPolicy {

  static let defaultRejectBelow = 40.0
  static let defaultAcceptAbove = 80.0

  let rejectBelow: Double
  let acceptAbove: Double

  /// Builds a validated policy. A non-finite input falls back to the per-field
  /// default; an INVERTED pair (`rejectBelow > acceptAbove`) falls back to BOTH
  /// defaults. `rejectBelow == acceptAbove` is allowed (empty WARN → binary).
  init(rejectBelow: Double? = nil, acceptAbove: Double? = nil) {
    let r = (rejectBelow.map { $0.isFinite } == true) ? rejectBelow! : Self.defaultRejectBelow
    let a = (acceptAbove.map { $0.isFinite } == true) ? acceptAbove! : Self.defaultAcceptAbove
    if r <= a {
      self.rejectBelow = r
      self.acceptAbove = a
    } else {
      self.rejectBelow = Self.defaultRejectBelow
      self.acceptAbove = Self.defaultAcceptAbove
    }
  }

  /// `score < rejectBelow` → reject; `score > acceptAbove` → accept; else warn
  /// (so exactly [rejectBelow] and [acceptAbove] are warn). Non-finite fails safe
  /// to reject — never accept.
  func classify(_ score: Double) -> BlurBand {
    if !score.isFinite { return .reject }
    if score < rejectBelow { return .reject }
    if score > acceptAbove { return .accept }
    return .warn
  }
}

File: android/app/src/main/kotlin/com/mayasabhaxr/recapture/camera/BlurThresholdPolicy.kt
// android/app/src/main/kotlin/com/mayasabhaxr/recapture/camera/BlurThresholdPolicy.kt
package com.mayasabhaxr.recapture.camera

/**
 * Three-band classification of a sharpness score (the variance-of-Laplacian from
 * [BlurMetric]) into an actionable band. The policy layer on top of the metric:
 * the blur-detection task computes the score, this turns it into a band the
 * capture flow + UI act on. It does NOT recompute the metric.
 *
 * The two thresholds (defaults 40 / 80) are content- and device-sensitive magic
 * numbers — configurable and intended to be remote-config-tunable without an app
 * release. The WARN band is an intentional buffer that avoids a hard sharp/blurry
 * flip at a single cutoff.
 */
enum class BlurBand {
    /** Too blurry — don't use the frame (discard / "hold steady" hint). */
    REJECT,

    /** Borderline — usable but flagged; the capture flow decides (accept-with-caution / prompt). */
    WARN,

    /** Sharp enough — use the frame. */
    ACCEPT;

    /** Lowercase wire form for the channel payload (`reject`/`warn`/`accept`). */
    val wire: String get() = name.lowercase()
}

/**
 * Holds the active (validated) thresholds and classifies scores against them.
 * Pure (no Android runtime), so it is JVM-unit-testable. Thread-safe for the
 * analyzer: thresholds are `@Volatile`, [update] applies to SUBSEQUENT
 * classifications (past frames are not reclassified).
 */
class BlurThresholdPolicy(
    rejectBelow: Double = DEFAULT_REJECT_BELOW,
    acceptAbove: Double = DEFAULT_ACCEPT_ABOVE,
) {
    companion object {
        const val DEFAULT_REJECT_BELOW = 40.0
        const val DEFAULT_ACCEPT_ABOVE = 80.0

        /**
         * Validates a threshold pair. A non-finite input falls back to the per-field
         * default; an INVERTED pair (`rejectBelow > acceptAbove`) is nonsensical, so
         * it falls back to BOTH defaults. `rejectBelow == acceptAbove` is allowed —
         * the WARN band is then empty (binary reject/accept).
         */
        fun validated(rejectBelow: Double?, acceptAbove: Double?): Pair<Double, Double> {
            val r = rejectBelow?.takeIf { it.isFinite() } ?: DEFAULT_REJECT_BELOW
            val a = acceptAbove?.takeIf { it.isFinite() } ?: DEFAULT_ACCEPT_ABOVE
            return if (r <= a) r to a else DEFAULT_REJECT_BELOW to DEFAULT_ACCEPT_ABOVE
        }
    }

    @Volatile
    var rejectBelow: Double = DEFAULT_REJECT_BELOW
        private set

    @Volatile
    var acceptAbove: Double = DEFAULT_ACCEPT_ABOVE
        private set

    init {
        update(rejectBelow, acceptAbove)
    }

    /** Applies validated thresholds; subsequent [classify] calls use the new values. */
    fun update(rejectBelow: Double?, acceptAbove: Double?) {
        val (r, a) = validated(rejectBelow, acceptAbove)
        this.rejectBelow = r
        this.acceptAbove = a
    }

    /**
     * Classifies a sharpness score. Boundaries (inclusive band ownership):
     * `score < rejectBelow` → REJECT; `score > acceptAbove` → ACCEPT; otherwise
     * WARN — so exactly [rejectBelow] and exactly [acceptAbove] are WARN. A
     * non-finite score (NaN/Inf) fails safe to REJECT — never ACCEPT.
     */
    fun classify(score: Double): BlurBand = when {
        !score.isFinite() -> BlurBand.REJECT
        score < rejectBelow -> BlurBand.REJECT
        score > acceptAbove -> BlurBand.ACCEPT
        else -> BlurBand.WARN
    }
}
